cmd/snap/cmd_run.go

```diff
@@ -333,7 +333,26 @@ func createUserDataDirs(info *snap.Info) error {
 		}
 	}
 
-	return createOrUpdateUserDataSymlink(info, usr)
+	if err := createOrUpdateUserDataSymlink(info, usr); err != nil {
+		return err
+	}
+
+	return maybeRestoreSecurityContext(filepath.Join(usr.HomeDir, dirs.UserHomeSnapDir))
```

Comment:
So does that mean we will re-label all of `~/snap` on *each* execution of a snap command?

Reply:
Unfortunately it does. Need to look into options of doing this bit smarter. We could just look at the top level label and run restorecon then. The caveat is the label is not easily readable with Go stdlib, we'd need to pull in `github.com/opencontainers/selinux/go-selinux`. OTOH, we may need to pull it in sooner or later anyway.

Reply:
To make discussing this branch more approachable can you add a comment that explains why we need to relabel $SNAP_USER_DATA on startup. Is $SNAP_DATA or other places also susceptible for similar treatment?

Reply:
The usual pattern is to have it done automatically with SELinux policy and proper file/directory transition (this is only implemented in #6238). However, that only works for newly created objects. Files or directories created before the new policy is introduced will be  incorrectly labeled as `user_home_t` instead of `snappy_home_t`.

Maybe we should only limit the relabeling to a case when `~/snap` exists and has incorrect label?

Reply:
Can we put the policy version into the system key and have snapd security backend perform the relabelling?

Reply:
I use the firefox snap, so I have 1.3G of ~/snap and quite a few files. Do we have a sense how long it takes that relabeling?

Reply:
Each file would need to be touched and updated. File count is relevant here. I've run some benchmarks in a VM:

These are small files:
```
[guest@localhost ~]$ find foo  | wc -l
592606
[guest@localhost ~]$ time restorecon -R foo

real    0m27.038s
user    0m15.923s
sys     0m5.575s
```

Something more realistic:
```
[guest@localhost ~]$ find foo | wc -l
10101
[guest@localhost ~]$ time restorecon -R foo

real    0m0.456s
user    0m0.305s
sys     0m0.119s
```

These are 1M files:
```
[guest@localhost ~]$ du -sch foo
1001M   foo
1001M   total
[guest@localhost ~]$ find foo  | wc -l
1101
[guest@localhost ~]$ time restorecon -R foo

real    0m0.404s
user    0m0.073s
sys     0m0.041s
````